### PR TITLE
sway-parse initial PR

### PR DESCRIPTION
**Pull request by canndrew**
This PR adds implements the new parser in a separate sway-parse crate in the cargo workspace.

You can currently switch between using the new parser or the old parser by modifying [this line](https://github.com/FuelLabs/sway/commit/ad209d69edbcc7ea30375c8ded640a30bf9d1c89#diff-066e2f8870fac537ada2033c79f060beaebb91c9d17af31fc413485add581deaR132) in sway-core/src/lib.rs. This should either be turned into a configurable option or removed entirely if/when we feel we're ready to commit to the new parser.

After the new parser gets run on a file its output is transformed into the old AST using the routines in sway-core/src/convert_parse_tree.rs. This is so that I didn't have to rewrite the typechecker to work on the new AST. However converting to the old AST also involves a lot desugaring/lowering so the old AST is now effectively a another intermediate representation.

### CODE REVIEW

**Comment by canndrew** [comment]
Something that I haven't done yet is add error recovery, so if there's any syntax errors then compilation will fail and just report the first error. I think, before adding that, we should write ui tests like the rust compiler has in order to specify what error output should look like and what errors should be recoverable.

**Review by adlerjohn — COMMENTED** [review]
ci failures

**Comment by canndrew** [comment]
CI should be fixed now.

File: Cargo.toml [hunk 1]
```diff
@@ -18,7 +17,8 @@ members = [
     "sway-utils",
     "test",
     "test-sig-gen-util",
-    "scripts/forc-documenter"
+    "scripts/forc-documenter",
+    "sway-parse",
```
Comment (adlerjohn) [inline Cargo.toml:21]: Alphabetize

File: sway-parse/src/expr/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1145 @@
+use crate::priv_prelude::*;
```
Comment (sezna) [inline sway-parse/src/expr/mod.rs:1]: I love the idea of a private prelude. We should use this more!

File: sway-parse/src/parser.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        Parser {
+            token_trees: token_stream.token_trees(),
+            errors,
+        }
+    }
+
+    pub fn emit_error(&mut self, kind: ParseErrorKind) -> ErrorEmitted {
+        let span = match self.token_trees.first() {
+            Some(token_tree) => token_tree.span(),
+            None => todo!(),
```
Comment (sezna) [inline sway-parse/src/parser.rs:19]: Preferably resolve this before merging

**Review by sezna — CHANGES_REQUESTED** [review]
This is awesome. I can only imagine how amazing it feels to be on the final stretches of this!

1. I think we should get this in ASAP before merge conflicts come in. As most of the team is in Amsterdam, this week should be a nice calm period, so hopefully we can manage this.
2. In a follow up PR, could you write some top-level module documentation strings describing the overall architecture and algorithm? I found most of the code easy to read and self-documenting, but some architectural-level descriptions from the person who wrote this would go a long way.
3. With the old AST being another IR....maybe @emilyaherbert can have fun with a four-stage pipeline like MIR, haha. This layer is definitely unnecessary but I agree that it's the only way to manage this new parser's integration for now.  Not the end of the world.

Could you change the new/old parser toggle to be a command line flag that defaults to the new parser? That way, similar to IR, we can do `--use-old-parser`? It can just be something `forc` passes in to `BuildConfig`

**Comment by eureka-cpu** [comment]
Is there a reference for issues that we were running into with the current parser that led to this PR (for tracking purposes)?

**Comment by sezna** [comment]
#252

File: sway-parse/src/parser.rs (hunk 3 above)
Comment (canndrew) [reply sway-parse/src/parser.rs:19]: Ooops. Good catch :grimacing:

File: sway-parse/src/expr/mod.rs (hunk 2 above)
Comment (canndrew) [reply sway-parse/src/expr/mod.rs:1]: Yeah I write all my crates like this. Other than making it easier to manage imports it also forces things to have a consistent meaning throughout the crate (eg. so `Span` doesn't mean two different things in two different files).

**Comment by canndrew** [comment]
> Could you change the new/old parser toggle to be a command line flag that defaults to the new parser? That way, similar to IR, we can do --use-old-parser? It can just be something forc passes in to BuildConfig

I've done this. There's now a `--use-orig-parser` option.

**Comment by canndrew** [comment]
This is now rebased to fix the conflicts.

**Comment by mitchmindtree** [comment]
Adding the new flag to the `forc` commands requires updating the `forc` docs. Unfortunately we don't have any way of generating these docs while running `mdbook build`, however we *do* have a new `forc-documenter` script that at least catches out-of-date docs during CI and that automates the process of updating the forc command docs. Could you run:

```
cargo run --bin forc-documenter write-docs
```

and then push a new commit with the result? That should update the `forc` docs and appease CI.

**Comment by adlerjohn** [comment]
Does this close #252?

**Comment by sezna** [comment]
> Does this close https://github.com/FuelLabs/sway/issues/252?
yes

**Comment by canndrew** [comment]
@sezna: I've made the requested changes. Is this ready to be merged?

**Review by sezna — APPROVED** [review]
Reviewing a PR this large is not super useful. I did scan it, and I also checked out the branch and tested it out. I'm sure there are things we are missing in the review, but it will probably make more sense to merge this in and handle the fallout, maybe holding a week or two before releasing so we can feel more solid about it.

**Pull request merged**